### PR TITLE
[Fluid] Process to apply boundary conditions according to Mach regime

### PR DESCRIPTION
**Pull request by EduardGomezEscandell**
**Description**
This PR aims to allow users to set boundary conditions depending on the mach regime of the flow. The options should mostly be the same as for other simple boundary condition processes (choice of variable, time interval and component when working with vectors).

The project parameters must specify two sets of boundary conditions, one for subsonic and one for supersonic. The process switches between one and the other on a per-node basis depending on the local mach number projected onto the boundary's normal.

Because the Mach number is scalar in Kratos, an auxiliary variable is used to know the direction of the flux (such as velocity or momentum, it can be chosen in ProjectParameters). The projected mach number is then:

proj(Ma) = abs(Ma * __v__·__n__ / |__v__|)

We use the absolute value because, for the moment, we consider that outlet/inlet discrimination must be done by the user when imposing the boundary conditions, rather than being detected on runtime.

**Changelog**
- Written C++ process ```ApplyMachDependentBoundaryConditions```
- Created Python wrapper ```apply_mach_depenedent_boundary_conditions.py```
- Created unit tests

### CODE REVIEW

File: applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.h [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+
+        /// Pointer definition of ApplyMachDependentBoundaryConditions
+        KRATOS_CLASS_POINTER_DEFINITION(ApplyMachDependentBoundaryConditions);
+
+        ///@}
+        ///@name Life Cycle
+        ///@{
+
+        /// Constructor
+        ApplyMachDependentBoundaryConditions(Model& rModel, Parameters& rParameters);
```
Comment (ddiezrod) [inline applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.h]: You do not need to pass Parameters as a reference. They are kind of pointers already, so you can send them by copy,

**Comment by RiccardoRossi** [comment]
actually passing them as reference can even be an error

File: applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.h (hunk 1 above)
Comment (EduardGomezEscandell) [reply applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.h]: True. Fixed now.

File: applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.h [hunk 2]
```diff
@@ -133,7 +133,7 @@ namespace Kratos
             /** @brief This object points to the proper (Fix|Free)Dof function
              * according to the time interval.
              */
-            std::function<decltype(FixDof)> mEnforceInternal = FreeDof;
+            decltype(FixDof) * mEnforceInternal = & FreeDof;
```
Comment (EduardGomezEscandell) [inline applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.h]: Apparently function pointers are considerably faster than std::function. Since I know I won't bind a lambda, I decided to switch to function pointer.

This function is assigned once per time-step and variable, and called once per time-step, variable, and node, so the gains may accumulate.

A ran a quick benchmark of a simplified version and apparently std::function is between 5 to 6 times slower.
OmBeMOJfzSKViWK9FIG2IRMdxpc

The benchmark: https://quick-bench.com/q/OmBeMOJfzSKViWK9FIG2IRMdxpc

File: applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+        // Chosing BC set to enforce according to mach
+        const bool supersonic = fabs(mach_projection) >= 1.0;
+        
+        const auto & active_bc  = supersonic ? mSupersonicBCs : mSubsonicBCs;
+        const auto & passive_bc = supersonic ? mSubsonicBCs   : mSupersonicBCs;
+
+        for(const auto & boundary_condition: passive_bc)
+        {
+            boundary_condition.Release(rNode);
+        }
```
Comment (EduardGomezEscandell) [inline applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp]: Just thought that it'll play nicer with other processes if the inactive BC's don't free their Dof here, and instead have all BC's free them during ```ExecuteFinalizeSolutionStep```, like similar processes do.

Should be fixed by tomorrow.

File: applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.h [hunk 4]
```diff
@@ -0,0 +1,314 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main author:     Eduard Gómez
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.h]: Header formatting.

File: applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.h [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+         * 
+         * This allows for checking the time only once at the beginning of the
+         * time-step rather than at every node.
+         */
+        class BoundaryConditionUtility
+        {
+        public:
+            BoundaryConditionUtility(const std::string & variable_name,
+                                     const double Value,
+                                     const IntervalUtility & rIntervalUtility);
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.h]:
```suggestion
            BoundaryConditionUtility(
                const std::string& rVariableName,
                const double Value,
                const IntervalUtility & rIntervalUtility);
```
Please follow the style guide convention (see [here](https://github.com/KratosMultiphysics/Kratos/wiki/Style-Guide))

File: applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.h [hunk 6]
```diff
@@ -0,0 +1,314 @@
...
+        ///@{
+
+
+        ///@}
+        ///@name Protected LifeCycle
+        ///@{
+
+
+        ///@}
+
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.h]: You can remove the `protected` section if not used.

File: applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.h [hunk 7]
```diff
@@ -0,0 +1,314 @@
...
+        ///@name Static Member Variables
+        ///@{
+
+
+        ///@}
+        ///@name Member Variables
+        ///@{
+
+        ModelPart * mpModelPart;
+        VectorVariable const * mpFlowDirectionVariable;
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.h]: I'd save a reference instead.

File: applications/FluidDynamicsApplication/tests/test_apply_mach_dependent_boundary_conditions.py [hunk 8]
```diff
@@ -0,0 +1,514 @@
+import KratosMultiphysics
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/tests/test_apply_mach_dependent_boundary_conditions.py]: Note that all the tests in the `FluidDynamicsApplication` have the keyword `test` at the end of the filename or class name. This is intentionally done to ease browsing the code. Please do it accordingly.

File: applications/FluidDynamicsApplication/tests/test_apply_mach_dependent_boundary_conditions.py [hunk 9]
```diff
@@ -0,0 +1,514 @@
...
+
+        # Subsonic case
+        for node in main_model_part.Nodes:
+            node.SetValue(FluidDynamicsApplication.MACH, 0.0)
+
+        process.ExecuteInitializeSolutionStep()
+
+        for node in main_model_part.Nodes:
+            self.assertTrue(node.IsFixed(KratosMultiphysics.DENSITY),
+                msg="Failed to fix subsonic boundary condition (Node #%d)." % node.Id)
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/tests/test_apply_mach_dependent_boundary_conditions.py]:
```suggestion
                msg="Failed to fix subsonic boundary condition (Node #{}).".format(node.Id))
```
We normally use `.format` rather than '%'.

File: applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp [hunk 10]
```diff
@@ -0,0 +1,296 @@
...
+    
+    mpFlowDirectionVariable = & KratosComponents<VectorVariable>::Get(flow_direction_name);
+
+    mRefreshNormalsEveryTimeStep = Parameters["refresh_normals_every_time_step"].GetBool();
+
+    // Reading boundary conditions to enforce
+    mSubsonicBCs.reserve(Parameters["subsonic_boundary_conditions"].size());
+    mSupersonicBCs.reserve(Parameters["supersonic_boundary_conditions"].size());
+
+    for(auto & var_settings : Parameters["subsonic_boundary_conditions"])
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp]:
```suggestion
    for(auto var_settings : Parameters["subsonic_boundary_conditions"])
```

File: applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp [hunk 11]
```diff
@@ -0,0 +1,296 @@
...
+    {
+        ReadBoundaryCondition(mSubsonicBCs, var_settings);
+    }
+
+    for(auto & var_settings : Parameters["supersonic_boundary_conditions"])
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp]:
```suggestion
    for(auto var_settings : Parameters["supersonic_boundary_conditions"])
```

File: applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp (hunk 10 above)
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp]: `Parameters["subsonic_boundary_conditions]` is of `KratosParameters` type.

File: applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp (hunk 11 above)
Comment (rubenzorrilla) [reply applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp]: Same comment.

File: applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp [hunk 12]
```diff
@@ -0,0 +1,296 @@
...
+    {
+        boundary_condition.ActivateIfInsideTimeInterval(time);
+    }
+
+    // Enforcing boundary conditions
+    block_for_each(mpModelPart->Nodes(), [&](NodeType & rNode)
+    {
+        // Computing mach projected onto normal
+        const auto & normal = rNode.FastGetSolutionStepValue(NORMAL);
+        const auto & flow_direction = rNode.FastGetSolutionStepValue(*mpFlowDirectionVariable);
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp]:
```suggestion
        const auto & r_normal = rNode.FastGetSolutionStepValue(NORMAL);
        const auto & r_flow_direction = rNode.FastGetSolutionStepValue(*mpFlowDirectionVariable);
```
Style guide.

File: applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp [hunk 13]
```diff
@@ -0,0 +1,296 @@
...
+
+        const double & mach = rNode.GetValue(MACH);
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp]:
```suggestion
        const double mach = rNode.GetValue(MACH);
```
Memory is the same so if it is not strictly required better use a copy.

File: applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp [hunk 14]
```diff
@@ -0,0 +1,296 @@
...
+        const double mach_projection = mach * inner_prod(normal, flow_direction) / norm_2(flow_direction);
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp]: You should check that `norm_2(flow_direction)` is non-zero.

File: applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp [hunk 15]
```diff
@@ -0,0 +1,296 @@
...
+        
+        // Chosing BC set to enforce according to mach
+        const bool supersonic = fabs(mach_projection) >= 1.0;
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp]:
```suggestion
        const bool supersonic = std::abs(mach_projection) >= 1.0;
```

File: applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp [hunk 16]
```diff
@@ -0,0 +1,296 @@
...
+        }
+    });
+}
+
+
+ApplyMachDependentBoundaryConditions::BoundaryConditionUtility::
+    BoundaryConditionUtility(const std::string & variable_name,
+                            const double Value,
+                            const IntervalUtility & rIntervalUtility)
+    : mValue(Value), mInterval(rIntervalUtility)
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp]:
```suggestion
ApplyMachDependentBoundaryConditions::BoundaryConditionUtility::BoundaryConditionUtility(
    const std::string &rVariableName,
    const double Value,
    const IntervalUtility& rIntervalUtility)
    : mValue(Value)
    , mInterval(rIntervalUtility)
```
Style guide.

File: applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp [hunk 17]
```diff
@@ -0,0 +1,296 @@
...
+        << "There is no double variable named " << variable_name << std::endl;
+
+    mpVariable = &KratosComponents<Variable<double>>::Get(variable_name);
+
+    KRATOS_CATCH("")
+}
+
+
+void ApplyMachDependentBoundaryConditions::BoundaryConditionUtility::
+    ActivateIfInsideTimeInterval(const double time)
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp]:
```suggestion
void ApplyMachDependentBoundaryConditions::BoundaryConditionUtility::ActivateIfInsideTimeInterval(const double Time)
```
Style guide (single argument functions in one line and CamelCase for arguments.

File: applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp [hunk 18]
```diff
@@ -0,0 +1,296 @@
...
+    }
+    else
+    {
+        mEnforceInternal = &DoNothing;
+    }
+}
+
+
+void ApplyMachDependentBoundaryConditions::BoundaryConditionUtility::
+    Enforce(NodeType & rNode) const
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp]:
```suggestion
void ApplyMachDependentBoundaryConditions::BoundaryConditionUtility::Enforce(NodeType& rNode) const
```

File: applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp [hunk 19]
```diff
@@ -0,0 +1,296 @@
...
+{
+    mEnforceInternal(*this, rNode);
+}
+
+
+void ApplyMachDependentBoundaryConditions::BoundaryConditionUtility::
+    Free(NodeType & rNode) const
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp]:
```suggestion
void ApplyMachDependentBoundaryConditions::BoundaryConditionUtility::Free(NodeType& rNode) const
```

File: applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp [hunk 20]
```diff
@@ -0,0 +1,296 @@
...
+            << "The number of values specified must be the same as number of constraints specified:\n"
+            << Parameters << std::endl;
+
+        KRATOS_ERROR_IF(n_components > 3)
+            << "Allowed vector variables are at most 3-dimensional:\n"
+            << Parameters << std::endl;
+
+        std::string variable_component_name = Parameters["variable_name"].GetString() + "_X";
+        
+        for(std::size_t i=0; i<n_components; i++)
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp]:
```suggestion
        for(std::size_t i=0; i<n_components; ++i)
```
Minor but this avoids a temporary copy of `i`. Hence, if not strictly needed to use `i++`, the `++i` version is preferred.

File: applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp [hunk 21]
```diff
@@ -0,0 +1,296 @@
...
+
+
+std::string ApplyMachDependentBoundaryConditions::Info() const
+{
+    return "ApplyMachDependentBoundaryConditions";
+}
+
+
+void ApplyMachDependentBoundaryConditions::
+    PrintInfo(std::ostream& rOStream) const
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp]:
```suggestion
void ApplyMachDependentBoundaryConditions::PrintInfo(std::ostream& rOStream) const
```

File: applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp [hunk 22]
```diff
@@ -0,0 +1,296 @@
...
+{
+    rOStream << "ApplyMachDependentBoundaryConditions";
+}
+
+
+void ApplyMachDependentBoundaryConditions::
+    PrintData(std::ostream& rOStream) const
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp]:
```suggestion
void ApplyMachDependentBoundaryConditions::PrintData(std::ostream& rOStream) const
```

File: applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp [hunk 23]
```diff
@@ -0,0 +1,296 @@
...
+{
+}
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp]: If there is no implementation I wouldn't override it.

File: applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.h [hunk 24]
```diff
@@ -0,0 +1,314 @@
...
+
+            /** @brief Does nothing. Used to replace FixDof when t is outside
+             * time interval
+             */
+            static void DoNothing(const BoundaryConditionUtility & rUtility, NodeType & rNode);
+
+            /** @brief This object points to the proper (FixDof|DoNothing)
+             * function according to the time interval.
+             */
+            decltype(FixDof) * mEnforceInternal = & DoNothing;
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.h]: I'd use an std::function pointer in here (type doesn't need to be inspected).

File: applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.h [hunk 25]
```diff
@@ -0,0 +1,314 @@
...
+            
+            /* @brief Fixes the Dof managed by this class and sets the value
+             * to the one specified by mValue
+             */
+            static void FixDof(const BoundaryConditionUtility & rUtility, NodeType & rNode);
+
+            /** @brief Does nothing. Used to replace FixDof when t is outside
+             * time interval
+             */
+            static void DoNothing(const BoundaryConditionUtility & rUtility, NodeType & rNode);
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.h]: I think we could avoid this (which is a bit ugly) by using an `std::function` pointer and checking if it is either `nullptr` or not.

File: applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp [hunk 26]
```diff
@@ -0,0 +1,296 @@
...
+
+
+void ApplyMachDependentBoundaryConditions::BoundaryConditionUtility::
+    Enforce(NodeType & rNode) const
+{
+    mEnforceInternal(*this, rNode);
+}
+
+
+void ApplyMachDependentBoundaryConditions::BoundaryConditionUtility::
```
Comment (rubenzorrilla) [inline applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp]: Why is this function needed? I have the feeling that it adds unnecessary complexity to the API.

**Review by rubenzorrilla — CHANGES_REQUESTED** [review]
Some minor comments:

- I'd call the class `ApplyCompressibleNavierStokesBoundaryConditionsProcess`. On the one hand this naming makes more clear which is the intended element (formulation) to be used in combination with it (the name modification should be propagated through the auxiliary classes and files). On the other hand it adds the `Process` keyword as we normally do in Kratos.
- Serialization is missing.
- A few years ago we had the discussion about attaching the `&` to the type or the declaration. For the sake of clarity (though behavior might be different in some cases) we decided to keep it attached to the type (e.g. `double& r_a`, not `double & r_a`). Although it might be annoying to change this, I'd appreciate sticking to this criterion in this class.
- I also have the feeling that we can simplify the design of the auxiliary class.

File: applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp (hunk 20 above)
Comment (EduardGomezEscandell) [reply applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp]: The compiler makes these equivalent even as low as `-O1`:
i++: https://godbolt.org/z/ccohxKqxn
++i: https://godbolt.org/z/njd53TMG8

there is no copy, they generate the same assembly

File: applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.h (hunk 24 above)
Comment (EduardGomezEscandell) [reply applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.h]: I tested it and function pointers are considerably faster. See https://github.com/KratosMultiphysics/Kratos/pull/9156#discussion_r717437479

File: applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.h (hunk 25 above)
Comment (EduardGomezEscandell) [reply applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.h]: Then we add a comparisson per node, boundary condition and time-step. The current implementation is uglier but removes these.

I can change it still if you believe the extra clarity is worth the performance hit.

File: applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp (hunk 26 above)
Comment (EduardGomezEscandell) [reply applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.cpp]: This was here to hide the internal storage of the utility (`mpVariable` is private). I can make it public if you think it's better.

**Comment by EduardGomezEscandell** [comment]
> * I'd call the class `ApplyCompressibleNavierStokesBoundaryConditionsProcess`. On the one hand this naming makes more clear which is the intended element (formulation) to be used in combination with it (the name modification should be propagated through the auxiliary classes and files). On the other hand it adds the `Process` keyword as we normally do in Kratos.

Agreed. Changed.

> * Serialization is missing.

Working on it.

> * A few years ago we had the discussion about attaching the `&` to the type or the declaration. For the sake of clarity (though behavior might be different in some cases) we decided to keep it attached to the type (e.g. `double& r_a`, not `double & r_a`). Although it might be annoying to change this, I'd appreciate sticking to this criterion in this class.

No problem. Changed.

> * I also have the feeling that we can simplify the design of the auxiliary class.

It is quite contrived because of the goal to minimize the number of comparissons. It can be simplified considerably if we're willing to sacrifice this.

File: applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.h (hunk 7 above)
Comment (EduardGomezEscandell) [reply applications/FluidDynamicsApplication/custom_processes/apply_mach_dependent_boundary_conditions.h]: For the modelpart it's simple enough, I can simply read the JSON before validating it.

But for the variable, it would entail calling KratosComponents::Get without validating the existance of the variable. This means that were the variable be misspelt, it would segfault without helping the user at all, at least when run not in debug mode. See https://github.com/KratosMultiphysics/Kratos/blob/c6b9da85068d55f8f6d9da01f9218f717d8bb1e0/kratos/includes/kratos_components.h#L96

**Review by rubenzorrilla — APPROVED** [review]
